src/core/init/stream_events_emitter/index.ts

```diff
@@ -0,0 +1,164 @@
+): evt is IStreamEventPayload {
+  return (evt as IStreamEventPayload).end !== undefined;
+}
+
+/**
+ * Tells if an event is included between two times.
+ * We don't know well calling this function which time
+ * is inferior to the other.
+ * @param {Object} event
+ * @param {number} timeA
+ * @param {number} timeB
+ */
+function isEventIncludedBetweenTimes(
+  event: IStreamEventPayload|IUnfiniteStreamEventPayload,
+  timeA: number,
+  timeB: number
+) {
+  const { start } = event;
+  const end = isFiniteStreamEvent(event) ? event.end : undefined;
+  const minTime = Math.min(timeA, timeB);
+  const maxTime = Math.max(timeA, timeB);
+  return minTime < start &&
+         maxTime >= (end ?? start);
+}
+
+/**
+ * Get events from manifest and emit each time an event has to be emitted
+ * @param {Object} manifest
+ * @param {HTMLMediaElement} mediaElement
+ * @returns {Observable}
+ */
+function streamEventsEmitter(manifest: Manifest,
+                             mediaElement: HTMLMediaElement,
+                             clock$: Observable<IInitClockTick>
+): Observable<IStreamEvent> {
+  const eventsBeingPlayed =
+    new WeakMap<IStreamEventPayload|IUnfiniteStreamEventPayload, true>();
+  const scheduledEvents$ = fromEvent(manifest, "manifestUpdate").pipe(
+    startWith(null),
+    scan((oldScheduleEvents) => {
+      return refreshScheduledEventsList(oldScheduleEvents, manifest);
+    }, [] as Array<IStreamEventPayload|IUnfiniteStreamEventPayload>)
+  );
+
+  return scheduledEvents$.pipe(switchMap(poll$));
+
+  function poll$(
+    newScheduleEvents: Array<IStreamEventPayload|IUnfiniteStreamEventPayload>
+  ) {
+    if (newScheduleEvents.length === 0) {
+      return EMPTY;
+    }
+    return observableCombineLatest([
+      interval(STREAM_EVENT_EMITTER_POLL_INTERVAL),
+      clock$,
+    ]).pipe(
+        map(([_, clockTick]) => {
+          const { seeking } = clockTick;
+          return { isSeeking: seeking,
+                   currentTime: mediaElement.currentTime };
+        }),
+        pairwise(),
+        mergeMap(([ oldTick, newTick ]) => {
+          const { isSeeking: wasSeeking, currentTime: previousTime } = oldTick;
+          const { isSeeking, currentTime } = newTick;
+          const eventsToSend: IStreamEvent[] = [];
+          for (let i = newScheduleEvents.length - 1; i >= 0; i--) {
+            const event = newScheduleEvents[i];
+            const start = event.start;
+            const end = isFiniteStreamEvent(event) ? event.end :
+                                                     undefined;
+            const isBeingPlayed = eventsBeingPlayed.has(event);
+            if (isBeingPlayed) {
+              if (start > currentTime ||
+                  (end !== undefined && currentTime >= end)
+              ) {
+                if (isFiniteStreamEvent(event) &&
+                    event.publicEvent.onExit !== undefined &&
+                    typeof event.publicEvent.onExit === "function") {
```

Comment:
Aren't the last two conditions redundant?

Reply:
Also there is a risk that a user call `rxPlayer.stop()` or something doing side-effects when we call `onExit`, in which case we might continue to send even more events after the `stop`.
Or worse, a user could dispose.

We have to find a way around that.